src/components/SingletonEditor.tsx

```diff
@@ -24,48 +31,97 @@ interface UseSingletonEditorConfig {
   onSave: (content: Content) => void;
 }
 
+interface UnlayerEditor extends EmailEditor {
+  removeEventListener: any;
+}
+
 export const SingletonDesignContext = createContext<ISingletonDesignContext>({
   hidden: true,
   setContent: () => {},
   editorState: { isLoaded: false, unlayer: undefined },
 });
 
-export const useSingletonEditor = ({
-  initialContent,
-  onSave,
-}: UseSingletonEditorConfig) => {
+export const useSingletonEditor = (
+  { initialContent, onSave }: UseSingletonEditorConfig,
+  deps: any[]
+) => {
   const { editorState, setContent } = useContext(SingletonDesignContext);
+  const hasChangesRef = useRef(false);
 
-  const saveHandler = () => {
-    if (!editorState.isLoaded) {
-      console.error("The editor is loading, can't save yet!");
-      return;
-    }
+  const saveHandler = useCallback(
+    (force?: boolean) => {
+      if (!hasChangesRef.current && !force) {
+        return;
+      }
+      if (!editorState.isLoaded) {
+        console.error("The editor is loading, can't save yet!");
+        return;
+      }
 
-    editorState.unlayer.exportHtml((htmlExport: HtmlExport) => {
-      const content = !htmlExport.design
-        ? {
-            htmlContent: htmlExport.html,
-            type: "html",
-          }
-        : {
-            design: htmlExport.design,
-            htmlContent: htmlExport.html,
-            type: "unlayer",
-          };
-
-      onSave(content as Content);
-    });
-  };
+      editorState.unlayer.exportHtml((htmlExport: HtmlExport) => {
+        const content = !htmlExport.design
+          ? {
+              htmlContent: htmlExport.html,
+              type: "html",
+            }
+          : {
+              design: htmlExport.design,
+              htmlContent: htmlExport.html,
+              type: "unlayer",
+            };
+
+        onSave(content as Content);
+      });
+    },
+    // eslint-disable-next-line
+    [editorState, ...deps]
+  );
 
   useEffect(() => {
+    hasChangesRef.current = false;
+    window.addEventListener("beforeunload", () => saveHandler());
+
+    const updateDesignListener = () => {
+      hasChangesRef.current = true;
+    };
+
+    const onLoadEventListener = () => {
+      editorState.unlayer &&
+        editorState.unlayer.addEventListener(
+          "design:updated",
+          updateDesignListener
+        );
+    };
+
+    editorState.isLoaded &&
+      editorState.unlayer.addEventListener(
+        "design:loaded",
+        onLoadEventListener
+      );
+
     setContent(initialContent);
+
     return () => {
+      window.removeEventListener("beforeunload", () => saveHandler());
```

Comment:
Mmm, creo que esto no va a funcionar porque entiendo que `() => saveHandler()` es una función diferente cada vez que se define.

En mi opinión, habría que volver atrás el cambio en el add

```diff
   useEffect(() => {
     hasChangesRef.current = false;
-     window.addEventListener("beforeunload", () => saveHandler());
+     window.addEventListener("beforeunload", saveHandler);

     const updateDesignListener = () => {
```

y acá también

```suggestion
      window.removeEventListener("beforeunload", saveHandler);
```

Reply:
Tienes razón, voy a hacer un cambio en el `saveHandler()` para poder lograr esto en la suscripción del evento.

Reply:
~~No creo que se necesite ningún cambio ya que el parámetro es opcional.~~

Ahh, los eventos siempre se llaman con un parámetro, bueno, hay varias formas de resolverlo, seguro encontrás alguna elegante.

Reply:
💪🏻

Reply:
Aquí lo apliqué no me parece muy elegante esta forma https://github.com/FromDoppler/doppler-editors-webapp/blob/869c3dfd2b5156c176151f0f60920b9a76b10ccf/src/components/SingletonEditor.tsx#L82